components/http-client/src/api_client.rs

```diff
@@ -191,121 +188,113 @@ impl ApiClient {
-///
-/// ```text
-/// <PRODUCT>/<VERSION> (<TARGET>; <KERNEL_RELEASE>)
-/// ```
-///
-/// where:
-///
-/// * `<PRODUCT>`: is the provided product name
-/// * `<VERSION>`: is the provided version string which may also include a release number
-/// * `<TARGET>`: is the machine architecture and the kernel separated by a dash in lower case
-/// * `<KERNEL_RELEASE>`: is the kernel release string from `uname`
-///
-/// For example:
-///
-/// ```text
-/// hab/0.6.0/20160606153031 (x86_64-darwin; 14.5.0)
-/// ```
-///
-/// # Errors
-///
-/// * If system information cannot be obtained via `uname`
-fn user_agent(product: &str, version: &str) -> Result<UserAgent> {
+fn user_agent(product: &str, version: &str) -> Result<HeaderValue> {
     let uname = sys::uname()?;
     let ua = format!("{}/{} ({}; {})",
                      product.trim(),
                      version.trim(),
                      PackageTarget::active_target(),
                      uname.release.trim().to_lowercase());
     debug!("User-Agent: {}", &ua);
-    Ok(UserAgent(ua))
+    Ok(HeaderValue::from_str(&ua).expect("Valid User Agent header"))
 }
 
-fn ssl_connector(fs_root_path: Option<&Path>) -> Result<SslConnector> {
-    let mut conn = SslConnectorBuilder::new(SslMethod::tls())?;
-    let mut options = SslOption::empty();
-    options.toggle(SSL_OP_NO_SSLV2);
-    options.toggle(SSL_OP_NO_SSLV3);
-    options.toggle(SSL_OP_NO_COMPRESSION);
-    ssl::set_ca(&mut conn, fs_root_path)?;
-    conn.set_options(options);
-    conn.set_cipher_list("ALL!EXPORT!EXPORT40!EXPORT56!aNULL!LOW!RC4@STRENGTH")?;
-
-    if env::var("HAB_SSL_CERT_VERIFY_NONE").is_ok() {
-        conn.set_verify(SSL_VERIFY_NONE);
+/// We need a set of root certificates when connected to SSL/TLS web endpoints.
+///
+/// The following strategy is used to locate a set of certificates that are used
+/// IN ADDITION to any system certificates that may be available (e.g., in /etc/ssl/certs or
+/// specified by a `SSL_CERT_FILE` environment variable):
+///
+/// 1. If the `core/cacerts` Habitat package is installed locally, then use the latest release's
+///    `cacert.pem` file.
+/// 2. If there is no 'core/cacerts packages, then a copy of `cacert.pem` will be written in an SSL
+///    cache directory (by default `/hab/cache/ssl` for a root user and `$HOME/.hab/cache/ssl` for
+///    a non-root user) and this will be used. The contents of this file will be inlined in this
+///    crate at build time as a fallback, which means that if the program using this code is
+///    operating in a minimal environment which may not contain any system certificates, it can
+///    still operate.
+/// 3. Other certs files (for example self-signed certs) that are found in the SSL cache directory
+///    will also get loaded into the root certs list. Both PEM and DER formats are supported (the
+///    extensions should be '.pem' or '.der' respectively)
+fn certificates(fs_root_path: Option<&Path>) -> Result<Vec<Certificate>> {
```

Comment:
This function is fairly complex. It would be good to get some test coverage on it. For that, though, it would probably be helpful to break it up into its 3 logical segments described in the doc comment.

Reply:
Simplified the function. For testing I'll open a separate issue to add cert cases to our e2e tests.